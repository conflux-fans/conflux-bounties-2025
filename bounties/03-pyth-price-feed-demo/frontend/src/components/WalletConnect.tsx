import React from 'react';
import { useAccount, useConnect, useDisconnect, useChainId, useSwitchChain } from 'wagmi';

export default function WalletConnect() {
  const { address, isConnected } = useAccount();
  const { connect, connectors } = useConnect();
  const { disconnect } = useDisconnect();
  const chainId = useChainId();
  const { switchChain } = useSwitchChain(); 

  const formatAddress = (addr: string) => {
    return `${addr.substring(0, 6)}...${addr.substring(addr.length - 4)}`;
  };

  if (isConnected && address) {
    return (
      <div className="flex items-center space-x-4">
        {chainId !== 1030 && (
          <button
            onClick={() => switchChain?.({ chainId: 1030 })} 
            className="px-4 py-2 bg-yellow-500 text-white rounded-lg hover:bg-yellow-600 transition-colors text-sm font-medium"
          >
            Switch to Conflux eSpace
          </button>
        )}
        
        <div className="flex items-center space-x-2 bg-gray-100 rounded-lg px-4 py-2">
          <div className="w-2 h-2 bg-green-500 rounded-full animate-pulse"></div>
          <span className="text-sm font-medium text-gray-700">
            {formatAddress(address)}
          </span>
        </div>

        <button
          onClick={() => disconnect()}
          className="px-4 py-2 bg-red-500 text-white rounded-lg hover:bg-red-600 transition-colors text-sm font-medium"
        >
          Disconnect
        </button>
      </div>
    );
  }

  return (
    <button
      onClick={() => connect({ connector: connectors[0] })}
      className="px-6 py-2 bg-gradient-to-r from-blue-500 to-indigo-600 text-white rounded-lg hover:from-blue-600 hover:to-indigo-700 transition-all shadow-md hover:shadow-lg text-sm font-medium"
    >
      Connect Wallet
    </button>
  );
}