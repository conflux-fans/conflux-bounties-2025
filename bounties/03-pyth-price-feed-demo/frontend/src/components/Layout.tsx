import React from 'react';
import WalletConnect from './WalletConnect';

interface LayoutProps {
  children: React.ReactNode;
}

export default function Layout({ children }: LayoutProps) {
  return (
    <div className="min-h-screen bg-gradient-to-br from-blue-50 to-indigo-100">
      <header className="bg-white shadow-md">
        <div className="container mx-auto px-4 py-4">
          <div className="flex items-center justify-between">
            <div className="flex items-center space-x-3">
              {/* Logo image */}
              <div className="w-10 h-10 rounded-lg overflow-hidden flex items-center justify-center bg-white">
                <img 
                  src="/confi.png" 
                  alt="Logo" 
                  className="w-full h-full object-contain"
                />
              </div>
              <div>
                <h1 className="text-2xl font-bold text-gray-900">
                  Pyth Price Feed
                </h1>
                <p className="text-sm text-gray-600">
                  Conflux eSpace Integration
                </p>
              </div>
            </div>
            <WalletConnect />
          </div>
        </div>
      </header>

      <main className="container mx-auto px-4 py-8">
        {children}
      </main>

      <footer className="bg-white mt-12 border-t border-gray-200">
        <div className="container mx-auto px-4 py-6">
          <div className="flex items-center justify-between text-sm text-gray-600">
            <div>
              <p>Pyth Price Feed Demo Dapp - Testing Purpose</p>
            </div>
            <div className="flex space-x-6">
              <a href="https://docs.pyth.network" target="_blank" rel="noopener noreferrer" className="hover:text-blue-600">
                Pyth Docs
              </a>
              <a href="https://doc.confluxnetwork.org" target="_blank" rel="noopener noreferrer" className="hover:text-blue-600">
                Conflux Docs
              </a>
              <a href="https://github.com/AmirMP12" target="_blank" rel="noopener noreferrer" className="hover:text-blue-600">
                GitHub
              </a>
            </div>
          </div>
        </div>
      </footer>
    </div>
  );
}