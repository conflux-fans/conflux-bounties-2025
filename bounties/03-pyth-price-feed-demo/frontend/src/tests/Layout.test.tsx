import React from 'react';
import { render, screen, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import Layout from '../components/Layout';
import WalletConnect from '../components/WalletConnect';

// Mock WalletConnect component
jest.mock('../components/WalletConnect', () => {
  return function MockWalletConnect() {
    return <div data-testid="wallet-connect">WalletConnect Mock</div>;
  };
});

describe('Layout Component', () => {
  const mockChildren = <div data-testid="test-child">Test Children Content</div>;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Component Structure', () => {
    test('renders the layout with all main sections', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      expect(screen.getByRole('banner')).toBeInTheDocument(); // header
      expect(screen.getByRole('main')).toBeInTheDocument();
      expect(screen.getByRole('contentinfo')).toBeInTheDocument(); // footer
    });

    test('applies correct background gradient classes', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const wrapper = container.firstChild as HTMLElement;
      expect(wrapper).toHaveClass('min-h-screen', 'bg-gradient-to-br', 'from-blue-50', 'to-indigo-100');
    });

    test('renders header with white background and shadow', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const header = screen.getByRole('banner');
      expect(header).toHaveClass('bg-white', 'shadow-md');
    });
  });

  describe('Header Section', () => {
    test('renders the logo image', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const logo = screen.getByRole('img', { name: /logo/i });
      expect(logo).toBeInTheDocument();
      expect(logo).toHaveAttribute('src', '/confi.png');
      expect(logo).toHaveAttribute('alt', 'Logo');
    });

    test('logo has correct styling classes', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const logo = screen.getByRole('img', { name: /logo/i });
      expect(logo).toHaveClass('w-full', 'h-full', 'object-contain');
    });

    test('logo container has correct dimensions', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const logoContainer = container.querySelector('.w-10.h-10.rounded-lg');
      expect(logoContainer).toBeInTheDocument();
    });

    test('renders main title', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      expect(screen.getByText('Pyth Price Feed')).toBeInTheDocument();
    });

    test('main title has correct styling', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const title = screen.getByText('Pyth Price Feed');
      expect(title).toHaveClass('text-2xl', 'font-bold', 'text-gray-900');
    });

    test('renders subtitle', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      expect(screen.getByText('Conflux eSpace Integration')).toBeInTheDocument();
    });

    test('subtitle has correct styling', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const subtitle = screen.getByText('Conflux eSpace Integration');
      expect(subtitle).toHaveClass('text-sm', 'text-gray-600');
    });

    test('renders WalletConnect component', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      expect(screen.getByTestId('wallet-connect')).toBeInTheDocument();
    });

    test('header content is properly aligned', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const headerContent = container.querySelector('.flex.items-center.justify-between');
      expect(headerContent).toBeInTheDocument();
    });

    test('logo and title are grouped together', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const logoTitleGroup = container.querySelector('.flex.items-center.space-x-3');
      expect(logoTitleGroup).toBeInTheDocument();
    });
  });

  describe('Main Content Section', () => {
    test('renders children content', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      expect(screen.getByTestId('test-child')).toBeInTheDocument();
      expect(screen.getByText('Test Children Content')).toBeInTheDocument();
    });

    test('children are rendered inside main element', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const main = screen.getByRole('main');
      const child = screen.getByTestId('test-child');
      
      expect(main).toContainElement(child);
    });

    test('main element has container and padding classes', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const main = screen.getByRole('main');
      expect(main).toHaveClass('container', 'mx-auto', 'px-4', 'py-8');
    });

    test('renders multiple children', () => {
      const multipleChildren = (
        <>
          <div data-testid="child-1">First Child</div>
          <div data-testid="child-2">Second Child</div>
          <div data-testid="child-3">Third Child</div>
        </>
      );

      render(<Layout>{multipleChildren}</Layout>);
      
      expect(screen.getByTestId('child-1')).toBeInTheDocument();
      expect(screen.getByTestId('child-2')).toBeInTheDocument();
      expect(screen.getByTestId('child-3')).toBeInTheDocument();
    });

    test('renders complex nested children', () => {
      const complexChildren = (
        <div>
          <h1>Page Title</h1>
          <section>
            <p>Content paragraph</p>
            <button>Action Button</button>
          </section>
        </div>
      );

      render(<Layout>{complexChildren}</Layout>);
      
      expect(screen.getByText('Page Title')).toBeInTheDocument();
      expect(screen.getByText('Content paragraph')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: /action button/i })).toBeInTheDocument();
    });

    test('renders when no children provided', () => {
      render(<Layout>{undefined}</Layout>);
      
      const main = screen.getByRole('main');
      expect(main).toBeInTheDocument();
      expect(main).toBeEmptyDOMElement();
    });
  });

  describe('Footer Section', () => {
    test('renders footer description', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      expect(screen.getByText('Pyth Price Feed Demo Dapp - Testing Purpose')).toBeInTheDocument();
    });

    test('footer has correct styling classes', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const footer = screen.getByRole('contentinfo');
      expect(footer).toHaveClass('bg-white', 'mt-12', 'border-t', 'border-gray-200');
    });

    test('footer content has correct layout', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const footerContent = container.querySelector('.flex.items-center.justify-between.text-sm.text-gray-600');
      expect(footerContent).toBeInTheDocument();
    });
  });

  describe('Footer Links', () => {
    test('renders Pyth Docs link', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const pythLink = screen.getByRole('link', { name: /pyth docs/i });
      expect(pythLink).toBeInTheDocument();
    });

    test('Pyth Docs link has correct attributes', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const pythLink = screen.getByRole('link', { name: /pyth docs/i });
      expect(pythLink).toHaveAttribute('href', 'https://docs.pyth.network');
      expect(pythLink).toHaveAttribute('target', '_blank');
      expect(pythLink).toHaveAttribute('rel', 'noopener noreferrer');
    });

    test('Pyth Docs link has hover styling', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const pythLink = screen.getByRole('link', { name: /pyth docs/i });
      expect(pythLink).toHaveClass('hover:text-blue-600');
    });

    test('renders Conflux Docs link', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const confluxLink = screen.getByRole('link', { name: /conflux docs/i });
      expect(confluxLink).toBeInTheDocument();
    });

    test('Conflux Docs link has correct attributes', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const confluxLink = screen.getByRole('link', { name: /conflux docs/i });
      expect(confluxLink).toHaveAttribute('href', 'https://doc.confluxnetwork.org');
      expect(confluxLink).toHaveAttribute('target', '_blank');
      expect(confluxLink).toHaveAttribute('rel', 'noopener noreferrer');
    });

    test('Conflux Docs link has hover styling', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const confluxLink = screen.getByRole('link', { name: /conflux docs/i });
      expect(confluxLink).toHaveClass('hover:text-blue-600');
    });

    test('renders GitHub link', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const githubLink = screen.getByRole('link', { name: /github/i });
      expect(githubLink).toBeInTheDocument();
    });

    test('GitHub link has correct attributes', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const githubLink = screen.getByRole('link', { name: /github/i });
      expect(githubLink).toHaveAttribute('href', 'https://github.com/AmirMP12');
      expect(githubLink).toHaveAttribute('target', '_blank');
      expect(githubLink).toHaveAttribute('rel', 'noopener noreferrer');
    });

    test('GitHub link has hover styling', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const githubLink = screen.getByRole('link', { name: /github/i });
      expect(githubLink).toHaveClass('hover:text-blue-600');
    });

    test('renders all three footer links', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const footer = screen.getByRole('contentinfo');
      const links = within(footer).getAllByRole('link');
      
      expect(links).toHaveLength(3);
    });

    test('footer links have correct spacing', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const linksContainer = container.querySelector('.flex.space-x-6');
      expect(linksContainer).toBeInTheDocument();
      
      const links = within(linksContainer as HTMLElement).getAllByRole('link');
      expect(links).toHaveLength(3);
    });

    test('all external links open in new tab securely', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const footer = screen.getByRole('contentinfo');
      const links = within(footer).getAllByRole('link');
      
      links.forEach(link => {
        expect(link).toHaveAttribute('target', '_blank');
        expect(link).toHaveAttribute('rel', 'noopener noreferrer');
      });
    });
  });

  describe('Responsive Layout', () => {
    test('header uses container class for responsive width', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const headerContainer = container.querySelector('header .container');
      expect(headerContainer).toBeInTheDocument();
      expect(headerContainer).toHaveClass('mx-auto', 'px-4', 'py-4');
    });

    test('main uses container class for responsive width', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const main = screen.getByRole('main');
      expect(main).toHaveClass('container', 'mx-auto');
    });

    test('footer uses container class for responsive width', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const footerContainer = container.querySelector('footer .container');
      expect(footerContainer).toBeInTheDocument();
      expect(footerContainer).toHaveClass('mx-auto', 'px-4', 'py-6');
    });
  });

  describe('Accessibility', () => {
    test('uses semantic HTML elements', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      expect(screen.getByRole('banner')).toBeInTheDocument(); // header
      expect(screen.getByRole('main')).toBeInTheDocument();
      expect(screen.getByRole('contentinfo')).toBeInTheDocument(); // footer
    });

    test('logo image has alt text', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const logo = screen.getByRole('img', { name: /logo/i });
      expect(logo).toHaveAttribute('alt', 'Logo');
    });

    test('all links are accessible', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const pythLink = screen.getByRole('link', { name: /pyth docs/i });
      const confluxLink = screen.getByRole('link', { name: /conflux docs/i });
      const githubLink = screen.getByRole('link', { name: /github/i });
      
      expect(pythLink).toBeInTheDocument();
      expect(confluxLink).toBeInTheDocument();
      expect(githubLink).toBeInTheDocument();
    });

    test('heading hierarchy is correct', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const heading = screen.getByRole('heading', { name: /pyth price feed/i });
      expect(heading.tagName).toBe('H1');
    });
  });

  describe('Visual Hierarchy', () => {
    test('title is styled as heading level 1', () => {
      render(<Layout>{mockChildren}</Layout>);
      
      const title = screen.getByRole('heading', { level: 1 });
      expect(title).toHaveTextContent('Pyth Price Feed');
    });

    test('footer text has proper text sizing', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const footerText = container.querySelector('.text-sm.text-gray-600');
      expect(footerText).toBeInTheDocument();
    });
  });

  describe('Integration', () => {
    test('WalletConnect is positioned correctly in header', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      const header = screen.getByRole('banner');
      const walletConnect = screen.getByTestId('wallet-connect');
      
      expect(header).toContainElement(walletConnect);
    });

    test('renders with real component structure', () => {
      const realPageContent = (
        <div className="page-content">
          <h2>Dashboard</h2>
          <div className="widgets">
            <div>Widget 1</div>
            <div>Widget 2</div>
          </div>
        </div>
      );

      render(<Layout>{realPageContent}</Layout>);
      
      expect(screen.getByText('Dashboard')).toBeInTheDocument();
      expect(screen.getByText('Widget 1')).toBeInTheDocument();
      expect(screen.getByText('Widget 2')).toBeInTheDocument();
    });
  });

  describe('Edge Cases', () => {
    test('handles null children gracefully', () => {
      const { container } = render(<Layout>{null}</Layout>);
      
      expect(container).toBeInTheDocument();
      expect(screen.getByRole('main')).toBeInTheDocument();
    });

    test('handles empty string children', () => {
      render(<Layout>{''}</Layout>);
      
      const main = screen.getByRole('main');
      expect(main).toBeInTheDocument();
    });

    test('handles React fragment as children', () => {
      render(
        <Layout>
          <>
            <div>Fragment Child 1</div>
            <div>Fragment Child 2</div>
          </>
        </Layout>
      );
      
      expect(screen.getByText('Fragment Child 1')).toBeInTheDocument();
      expect(screen.getByText('Fragment Child 2')).toBeInTheDocument();
    });

    test('handles conditional children rendering', () => {
      const showContent = true;
      
      render(
        <Layout>
          {showContent && <div>Conditional Content</div>}
        </Layout>
      );
      
      expect(screen.getByText('Conditional Content')).toBeInTheDocument();
    });
  });

  describe('Snapshot Testing', () => {
    test('matches snapshot with basic children', () => {
      const { container } = render(<Layout>{mockChildren}</Layout>);
      
      expect(container.firstChild).toMatchSnapshot();
    });

    test('matches snapshot with complex children', () => {
      const complexChildren = (
        <div>
          <h1>Test Page</h1>
          <p>Test content</p>
        </div>
      );

      const { container } = render(<Layout>{complexChildren}</Layout>);
      
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});