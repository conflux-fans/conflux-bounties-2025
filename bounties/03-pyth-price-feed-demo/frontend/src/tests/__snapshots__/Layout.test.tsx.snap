// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Component Snapshot Testing matches snapshot with basic children 1`] = `
<div
  class="min-h-screen bg-gradient-to-br from-blue-50 to-indigo-100"
>
  <header
    class="bg-white shadow-md"
  >
    <div
      class="container mx-auto px-4 py-4"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center space-x-3"
        >
          <div
            class="w-10 h-10 rounded-lg overflow-hidden flex items-center justify-center bg-white"
          >
            <img
              alt="Logo"
              class="w-full h-full object-contain"
              src="/confi.png"
            />
          </div>
          <div>
            <h1
              class="text-2xl font-bold text-gray-900"
            >
              Pyth Price Feed
            </h1>
            <p
              class="text-sm text-gray-600"
            >
              Conflux eSpace Integration
            </p>
          </div>
        </div>
        <div
          data-testid="wallet-connect"
        >
          WalletConnect Mock
        </div>
      </div>
    </div>
  </header>
  <main
    class="container mx-auto px-4 py-8"
  >
    <div
      data-testid="test-child"
    >
      Test Children Content
    </div>
  </main>
  <footer
    class="bg-white mt-12 border-t border-gray-200"
  >
    <div
      class="container mx-auto px-4 py-6"
    >
      <div
        class="flex items-center justify-between text-sm text-gray-600"
      >
        <div>
          <p>
            Pyth Price Feed Demo Dapp - Testing Purpose
          </p>
        </div>
        <div
          class="flex space-x-6"
        >
          <a
            class="hover:text-blue-600"
            href="https://docs.pyth.network"
            rel="noopener noreferrer"
            target="_blank"
          >
            Pyth Docs
          </a>
          <a
            class="hover:text-blue-600"
            href="https://doc.confluxnetwork.org"
            rel="noopener noreferrer"
            target="_blank"
          >
            Conflux Docs
          </a>
          <a
            class="hover:text-blue-600"
            href="https://github.com/AmirMP12"
            rel="noopener noreferrer"
            target="_blank"
          >
            GitHub
          </a>
        </div>
      </div>
    </div>
  </footer>
</div>
`;

exports[`Layout Component Snapshot Testing matches snapshot with complex children 1`] = `
<div
  class="min-h-screen bg-gradient-to-br from-blue-50 to-indigo-100"
>
  <header
    class="bg-white shadow-md"
  >
    <div
      class="container mx-auto px-4 py-4"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center space-x-3"
        >
          <div
            class="w-10 h-10 rounded-lg overflow-hidden flex items-center justify-center bg-white"
          >
            <img
              alt="Logo"
              class="w-full h-full object-contain"
              src="/confi.png"
            />
          </div>
          <div>
            <h1
              class="text-2xl font-bold text-gray-900"
            >
              Pyth Price Feed
            </h1>
            <p
              class="text-sm text-gray-600"
            >
              Conflux eSpace Integration
            </p>
          </div>
        </div>
        <div
          data-testid="wallet-connect"
        >
          WalletConnect Mock
        </div>
      </div>
    </div>
  </header>
  <main
    class="container mx-auto px-4 py-8"
  >
    <div>
      <h1>
        Test Page
      </h1>
      <p>
        Test content
      </p>
    </div>
  </main>
  <footer
    class="bg-white mt-12 border-t border-gray-200"
  >
    <div
      class="container mx-auto px-4 py-6"
    >
      <div
        class="flex items-center justify-between text-sm text-gray-600"
      >
        <div>
          <p>
            Pyth Price Feed Demo Dapp - Testing Purpose
          </p>
        </div>
        <div
          class="flex space-x-6"
        >
          <a
            class="hover:text-blue-600"
            href="https://docs.pyth.network"
            rel="noopener noreferrer"
            target="_blank"
          >
            Pyth Docs
          </a>
          <a
            class="hover:text-blue-600"
            href="https://doc.confluxnetwork.org"
            rel="noopener noreferrer"
            target="_blank"
          >
            Conflux Docs
          </a>
          <a
            class="hover:text-blue-600"
            href="https://github.com/AmirMP12"
            rel="noopener noreferrer"
            target="_blank"
          >
            GitHub
          </a>
        </div>
      </div>
    </div>
  </footer>
</div>
`;
